src/main/kotlin/org/rust/ide/newProject/RsDirectoryProjectGenerator.kt

```diff
@@ -35,28 +36,45 @@ class RsDirectoryProjectGenerator : DirectoryProjectGeneratorBase<ConfigurationD
 
     override fun validate(baseDirPath: String): ValidationResult {
         val crateName = File(baseDirPath).nameWithoutExtension
-        val isBinary = peer?.settings?.createBinary == true
-        val message = RsPackageNameValidator.validate(crateName, isBinary) ?: return ValidationResult.OK
+        val message = peer?.settings?.template?.validateProjectName(crateName) ?: return ValidationResult.OK
         return ValidationResult(message)
     }
 
     override fun generateProject(project: Project, baseDir: VirtualFile, data: ConfigurationData, module: Module) {
-        val (settings, createBinary) = data
+        val (settings, template) = data
+        val cargo = settings.toolchain?.rawCargo() ?: return
+
         val generatedFiles = project.computeWithCancelableProgress("Generating Cargo project...") {
-            settings.toolchain?.rawCargo()?.init(project, module, baseDir, createBinary)
+            makeProject(project, module, baseDir, cargo, template)
         } ?: return
 
-        // Open new files
-        if (!isHeadlessEnvironment) {
-            val navigation = PsiNavigationSupport.getInstance()
-            navigation.createNavigatable(project, generatedFiles.manifest, -1).navigate(false)
-            for (file in generatedFiles.sourceFiles) {
-                navigation.createNavigatable(project, file, -1).navigate(true)
-            }
-        }
+        openFiles(project, generatedFiles)
     }
 
     override fun createStep(projectGenerator: DirectoryProjectGenerator<ConfigurationData>,
                             callback: AbstractNewProjectStep.AbstractCallback<ConfigurationData>): AbstractActionWithPanel =
         RsProjectSettingsStep(projectGenerator)
 }
+
+fun makeProject(
```

Comment:
Maybe it's worth to convert this function to extension method of `Cargo` not to put quite specific function into global namespace?